keps/sig-scheduling/20180409-scheduling-framework.md

```diff
@@ -176,8 +176,15 @@ plugin may be enabled at a time.
 ### Pre-filter
 
 These plugins are used to pre-process info about the pod, or to check certain
-conditions that the cluster or the pod must meet. If a pre-filter plugin returns
-an error, the scheduling cycle is aborted.
+conditions that the cluster or the pod must meet. A pre-filter plugin should implement
+a PreFilter function, if PreFilter returns an error, the scheduling cycle is aborted.
+Note that PreFilter is called once in each scheduling cycle.
+
+A Pre-filter plugin can provide another two optional functions: **AddPod** and **RemovePod** 
+to incrementally modify its pre-processed info. The framework guarantees that those
+functions will only be called after PreFilter. The framework may call those functions 
+more than once before calling Filter on a specific node.
```

Comment:
We can mention that "Under the hood, when the framework call AddPod() and RemovePod(), it performs on a cloned object".

Reply:
oh, right, we need to also make a copy of the context. We should change the [ContextData](https://github.com/kubernetes/kubernetes/blob/a2760c9c139d8f3efa6eaeb02a553c470362e8e5/pkg/scheduler/framework/v1alpha1/context.go#L30) interface to add a Clone method.